airflow/providers/cncf/kubernetes/operators/spark_kubernetes.py

```diff
@@ -32,47 +52,283 @@ class SparkKubernetesOperator(BaseOperator):
+            pod = pod_list[0]
+            self.log.info(
+                "Found matching driver pod %s with labels %s", pod.metadata.name, pod.metadata.labels
+            )
+            self.log.info("`try_number` of task_instance: %s", context['ti'].try_number)
+            self.log.info("`try_number` of pod: %s", pod.metadata.labels['try_number'])
+        return pod
+
+    def get_or_create_spark_crd(self, launcher: CustomObjectLauncher, context) -> k8s.V1Pod:
+        if self.reattach_on_restart:
+            driver_pod = self.find_spark_job(context)
+            if driver_pod:
+                return driver_pod
+
+        # self.log.debug("Starting spark job:\n%s", yaml.safe_dump(launcher.body.to_dict()))
+        driver_pod, spark_obj_spec = launcher.start_spark_job(
+            image=self.image,
+            code_path=self.code_path,
+            startup_timeout=self.startup_timeout_seconds
+        )
+        return driver_pod
+
+    def extract_xcom(self, pod):
+        """Retrieves xcom value and kills xcom sidecar container"""
+        result = self.pod_manager.extract_xcom(pod)
+        self.log.info("xcom result: \n%s", result)
+        return json.loads(result)
+
+    def cleanup(self, pod: k8s.V1Pod, remote_pod: k8s.V1Pod):
+        pod_phase = remote_pod.status.phase if hasattr(remote_pod, 'status') else None
+        if not self.delete_on_termination:
+            with _suppress(Exception):
+                self.patch_already_checked(pod)
+        if pod_phase != PodPhase.SUCCEEDED:
+            if self.log_events_on_failure:
+                with _suppress(Exception):
+                    for event in self.pod_manager.read_pod_events(pod).items:
+                        self.log.error("Pod Event: %s - %s", event.reason, event.message)
+            with _suppress(Exception):
+                self.process_spark_job_deletion(pod)
+        else:
+            with _suppress(Exception):
+                self.process_spark_job_deletion(pod)
+
+    def process_spark_job_deletion(self, pod):
+        if self.delete_on_termination:
+            self.log.info("Deleting spark job: %s", pod.metadata)
+            self.launcher.delete_spark_job(pod.metadata.name.replace('-driver', ''))
+        else:
+            self.log.info("skipping deleting spark job: %s", pod.metadata.name)
+
+    @cached_property
+    def hook(self) -> KubernetesHook:
+        hook = KubernetesHook(
+            conn_id=self.kubernetes_conn_id,
+            in_cluster=self.template_body['kubernetes']['in_cluster'],
+            config_file=self.template_body['kubernetes']['kube_config_file'],
+            cluster_context=self.cluster_context,
         )
-        return response
+        return hook
+
+    @cached_property
+    def client(self) -> CoreV1Api:
+        return self.hook.core_v1_client
+
+    @cached_property
+    def custom_obj_api(self) -> CustomObjectsApi:
+        return CustomObjectsApi()
+
+    def execute(self, context: Context):
```

Comment:
You are duplicating a lot of code from KPO.  Do you think it's feasible to instead subclass it?

Reply:
It seems maybe you would not even need to change `execute` method but could simply override `build_pod_request_obj` -- does that sound right?

Reply:
And then it could be something like 

```python
def execute():
    self.create_crd()
    return super().execute()
```

that sound right?

Reply:
I did make a change to use it as a subclass. I had to change slightly the KPO but it is working now. https://github.com/apache/airflow/pull/22253/files#diff-947f9f98b4aedccee6dd81558568c350c738f5da050aad1224d8b48ec43677cdL437

what do you think?

Reply:
hey, great, let me take a look

Reply:
@dstandish did you get a chance to take a look?
Once you give me the green light I will revisit the tests.
We have started to use this version since last week.